src/kernel/arch/aarch64/tty.zig

```diff
@@ -2,93 +2,120 @@ const std = @import("std");
-    log.logDebug("Writing pixel {} to ({}, {}) at fb {}, which is address {x}\n", .{ pixel, x, y, @ptrToInt(framebuffer.buffer), @ptrToInt(&framebuffer.buffer[y * framebuffer.columns + x]) });
-    framebuffer.buffer[y * framebuffer.columns + x] = pixel;
+    framebuffer.buffer[y * framebuffer.bytes_per_row / 4 + x] = pixel;
+}
+
+fn clearScreen() void {
+    var y: usize = 0;
+    while (y < framebuffer.height) : (y += 1) {
+        var x: usize = 0;
+        while (x < framebuffer.width) : (x += 1) {
+            writePixel(x, y, BLACK);
+        }
+    }
 }
 
 fn writeChar(x: usize, y: usize, char: u8) void {
     var ch = char;
-    if (char < ' ' or char > '~')
+    if (ch < ' ' or ch > '~')
         ch = ' ';
-    const bitmap = font[ch - ' '];
-    var x2 = x;
-    var y2 = y;
-    for (bitmap) |bit, i| {
-        const pixel = if (bit == 0) BLACK else WHITE;
-        writePixel(x2, y2, pixel);
-        x2 += 1;
-        if (x2 >= CHAR_WIDTH) {
-            x2 = x;
-            y2 += 1;
+    var font_index = ch - ' ';
+    if (font_index >= font.len) {
+        font_index = 1;
+    }
+    const bitmap = font[font_index];
+    const left = x * CHAR_WIDTH;
+    const top = y * CHAR_HEIGHT;
+    var pixel_x = left;
+    var pixel_y = top;
+    for (bitmap) |bit| {
+        writePixel(pixel_x, pixel_y, if (bit == 0) BLACK else WHITE);
+        pixel_x += 1;
+        if (pixel_x == left + CHAR_WIDTH) {
+            pixel_x = left;
+            pixel_y += 1;
         }
     }
 }
 
-fn writeString(str: []const u8) !void {
+pub fn writeString(str: []const u8) !void {
     for (str) |ch| {
         if (ch == '\n') {
-            setCursor(0, framebuffer.y + 1);
+            setCursor(0, framebuffer.text_cursor_y + 1);
         } else {
-            if (framebuffer.y < framebuffer.rows and framebuffer.x < framebuffer.columns) {
-                writeChar(framebuffer.x, framebuffer.y, ch);
-                setCursor(framebuffer.x + 1, framebuffer.y);
-                if (framebuffer.x >= framebuffer.columns) {
-                    setCursor(0, framebuffer.y + 1);
+            if (framebuffer.text_cursor_y < framebuffer.text_rows and framebuffer.text_cursor_x < framebuffer.text_columns) {
+                writeChar(framebuffer.text_cursor_x, framebuffer.text_cursor_y, ch);
+                setCursor(framebuffer.text_cursor_x + 1, framebuffer.text_cursor_y);
+                if (framebuffer.text_cursor_x >= framebuffer.text_columns) {
+                    setCursor(0, framebuffer.text_cursor_y + 1);
                 }
             }
         }
     }
 }
 
-fn setCursor(x: u8, y: u8) void {
-    framebuffer.x = x;
-    framebuffer.y = y;
+pub fn setCursor(x: u8, y: u8) void {
+    framebuffer.text_cursor_x = x;
+    framebuffer.text_cursor_y = y;
 }
 
-pub fn init(allocator2: *std.mem.Allocator, board: arch.BootPayload) TTY {
+pub fn init(allocator2: *std.heap.FixedBufferAllocator, board: arch.BootPayload) TTY {
```

Comment:
Why is this using a specific allocator type? It's better for flexibility if the implementation doesn't care what type of allocator is used.

Reply:
Done